kernel/work.c

```diff
@@ -653,6 +653,12 @@ static void work_queue_main(void *workq_ptr, void *p2, void *p3)
 			 * submissions.
 			 */
 			(void)z_sched_wake_all(&queue->drainq, 1, NULL);
+		} else if (flag_test(&queue->flags, K_WORK_QUEUE_STOP_BIT)) {
+			/* User has requested that the queue stop. Clear the status flags and exit.
+			 */
+			flags_set(&queue->flags, 0);
+			k_spin_unlock(&lock, key);
+			return;
```

Comment:
Pardon me.
If we allow task append to the queue after setting `K_WORK_QUEUE_STOP_BIT`, the block will never be executed.

Reply:
A pre-condition to the _stop(..) function is that the queue has already been plugged. 

```
static inline int queue_submit_locked(struct k_work_q *queue,
				      struct k_work *work)
{
	 * * -ENODEV if the queue isn't running.
	 * * -EBUSY if draining and not chained
	 * * -EBUSY if plugged and not draining
	 * * otherwise OK

```

So the case you're thinking of is: 

Thread a. Continously adds work to queue. 

Thread b. pluggs queue & tries to stop the queue with K_FOREVER. 

Would thread b. freeze indefenately? 
Initially, maybe :+1: 

Question is if we should disallow that append of work if the K_WORK_QUEU_STOP_BIT is set as well. 
I'm not sure. How do you reason?

Reply:
Oh, I missed we need to call `k_work_queue_drain` before stopping `workqueue`, but the design make me thinks about the scenario will happen?
```
Thread a {
          while (true) {
                   queue_submit_locked(....);
         }
}

Thread b    k_work_queue_drain(..., true);
Thread c    k_work_queue_drain(..., true);
Thread b    k_work_queue_stop(..., K_FOREVER);
Thread c    k_work_queue_unplug(...);
```

Reply:
Sounds like a great opportunity for additional test cases

Reply:
@TaiJuWu 

Another option is to forbid the unplug function if `K_WORK_QUEUE_STOP_BIT` is set. 
```
Thread a {
          while (true) {
                   queue_submit_locked(....); <- Error once K_WORK_QUEUE_PLUGGED_BIT is set
         }
}

Thread b    k_work_queue_drain(..., true);
Thread c    k_work_queue_drain(..., true);
Thread b    k_work_queue_stop(..., K_FOREVER); <- Would exit in finite time
Thread c    k_work_queue_unplug(...); <- Would yield error & not unplug. 
```
Either thread B or C would return an error, depending on which accessed the queue first.
I believe this is the smallest change we could make to implement this feature. It shouldn't affect current use cases, what do you think? 

@cfriedt Agreed!

Reply:
The case I'm making could still deadlock thread b, if the queue never drains the queue will continue to accept works. 

So we probbably need the 2. You outlined earlier as well.
> 2. Disallow append if `K_WORK_QUEUE_STOP_BIT` is set.

But that changes the behaviour of the queue & drain system slightly, is that ok?

Reply:
> Another option is to forbid the unplug function if K_WORK_QUEUE_STOP_BIT is set.

Sound great.

>But that changes the behaviour of the queue & drain system slightly, is that ok?

It is ok for me but we still need @andyross 's or other maintainers comfirm.

Reply:
> By the way, I am not sure we need to call k_work_queue_drain before k_work_queue_stop by users is good api design or not.

Myself, being biased (haha), think it is neat :P The `queue_drain(..., true) ` gives the user the knowledge that the works scheduled up to that point will be executed and in so doing letting free-routines kick in as to not casue memory leaks in the application. 
`The k_work_queue_stop(...)` stopps the thread as soon as the last job has been processed. 

I'll commit the changes that we outlined here today :)

Reply:
I think we need to call `queue_drain(..., true)` in `k_work_queue_stop` and set `K_WORK_QUEUE_STOP_BIT ` berfore `queue_drain(..., true)`
Is it possible?
```
TA                                     TB
queue_drain(..., true)
                                    k_work_queue_stop but flag is not set
queue_unplug
                                   flag is set

```

Reply:
> Additionally, [System Workqueue](https://docs.zephyrproject.org/latest/kernel/services/threads/workqueue.html#id4)[](https://docs.zephyrproject.org/latest/kernel/services/threads/workqueue.html#system-workqueue) should never stop I guess. 

The patch specifically addresses scenarios where stopping a workqueue becomes interesting, like in a couple of tests I have in mind. 
I believe the patch aligns with the broader system goals. If this approach conflicts with project guidelines or expectations, I expect that the maintainers wil let me know shortly haha :smiling_face_with_tear:  

> I think we need to call queue_drain(..., true) in k_work_queue_stop internally and set K_WORK_QUEUE_STOP_BIT  berfore queue_drain(..., true)
Is it possible in current design? 

I've just looked through the `kernel/work.c` file, it would be a neat addition, one less function call from the users to achive the same effect. What's stopping me is the `k_work_queue_drain(...)` since the current implementation is blocking ( `K_FOREVER` ) and to resolve that I would need to rework that function to include a timeout and then manage to make the timeout time be split between the draining & the stopping of the thread (the stopping of the queue would be, almost, instantanious at that point if we've passed the drain stage). 
@cfriedt @andyross How would you like me to proceede with: 

-  to sort out the`_drain(..)` to include a timeout and incorperate parts of that function in the `_stop(...)` 

or

- keep it as is with the addition of blocking `_unplug(..)` and stop new works to be scheduled if the stop bit is set? 

Regarding the race-condition you're (@TaiJuWu) outlining I don't think it is a problem since the getting & setting of the flags are protected by the spin_lock, the winner of the spin_lock would dictate the outcome.